# Solving Systems of Linear Equations with NumPy - Lab

## Introduction 

Now we have gathered all the required skills needed to solve systems of linear equations. We saw why need to calculate inverses of matrices, followed by matrix multiplication to figure the values of unknown variables. 

Following exercises present some problems that can be converted into a system of linear equations. Solve the problems for the values of unknown variables. 

## Objectives
You will be able to:

* Build a system of equations from given problem by creating matrix and vectors in NumPy
* Calculate the inverse and dot-product to get the values for unknown variables.
* Solve linear equations using `numpy.linalg.solve()`

## Exercise 1

A coffee shop is having a sale on coffee and tea. 

On day 1, 29 bags of coffee and 41 bags of tea was sold, giving a total of 490 dollars.

On day 2, they sold 23 bags of coffee and 41 bags of tea, for which customers paid a total of 448 dollars.  

How much does each bag cost?

#### Equations:
* D1 => 29c + 41t = 490
* D2 => 23c + 41t = 448

In [1]:
import numpy as np

In [9]:
# define A and B
A = np.matrix([[29, 41],[23, 41]])
B = np.matrix([490, 448])
print(A)
print(B)

[[29 41]
 [23 41]]
[[490 448]]


In [10]:
#Take the inverse of A
Ainv = np.linalg.inv(A)
Ainv

matrix([[ 0.16666667, -0.16666667],
        [-0.09349593,  0.11788618]])

In [11]:
#check for shape of B before transposing
print (B.shape)

(1, 2)


In [12]:
#transpose B
B = B.T
B

matrix([[490],
        [448]])

In [13]:
print(B.shape)

(2, 1)


In [14]:
# calculate c (coffee) and t (tea) values
X = Ainv.dot(B)
X

matrix([[7.],
        [7.]])

The price for each bag of coffee or tea is $7.0

In [15]:

print(A.dot(X))
print(B)

[[490.]
 [448.]]
[[490]
 [448]]


## Exercise 2:

The cost of admission to a popular music concert was 162 dollars for 12 children and 3 adults. 

The admission was 122 dollars for 8 children and 3 adults in another music concert. 

How much was the admission for each child and adult?

#### Equations:
* 12c + 3a = 162
* 8c + 3a = 122

In [16]:
# Create a system of equations

A = np.matrix([[12, 3],[8,3]])
B = np.matrix([162, 122])

print(A)
print(B)

[[12  3]
 [ 8  3]]
[[162 122]]


In [17]:
Ainv = np.linalg.inv(A)
Ainv

matrix([[ 0.25      , -0.25      ],
        [-0.66666667,  1.        ]])

In [18]:
print (B.shape)

(1, 2)


In [19]:
B = B.T
print(B)
print(B.shape)

[[162]
 [122]]
(2, 1)


In [20]:
X = Ainv.dot(B)
X

matrix([[10.],
        [14.]])

* Admission for each child 10.0 dollars
* Admission for each adult 14.0 dollars

In [21]:

print(A.dot(X))
print(B)

[[162.]
 [122.]]
[[162]
 [122]]


## Exercise 3:

You want to make a certain kind of tropical punch, using bananas, oranges, and papayas. 

Suppose you don't know how many of each to put in the punch, but you know that there are seven pieces of fruit in the mix, and there are twice as many oranges as bananas. 

You also know that the seven pieces of fruit cost \$5.25, 

where bananas cost \$.50 each, oranges cost \$.75 each, and papayas cost \$1.25 each.

How many bananas, oranges and papayas would you need to make this punch ?

#### Equations:
* b + o + p = 7
* 1o = 2b => -2b + 1o + 0p = 0
* 0.50b + 0.75o + 1.25p = 5.25

In [27]:
# Create a system of equations
A = np.matrix([[1,1,1],[-2,1,0],[0.50,0.75,1.25]])
B = np.matrix([7, 0, 5.25])

print(A)
print(B)

[[ 1.    1.    1.  ]
 [-2.    1.    0.  ]
 [ 0.5   0.75  1.25]]
[[7.   0.   5.25]]


In [28]:
Ainv = np.linalg.inv(A)
Ainv

matrix([[ 0.71428571, -0.28571429, -0.57142857],
        [ 1.42857143,  0.42857143, -1.14285714],
        [-1.14285714, -0.14285714,  1.71428571]])

In [29]:
B = B.T
print (B)
print(B.shape)

[[7.  ]
 [0.  ]
 [5.25]]
(3, 1)


In [30]:
X = Ainv.dot(B)
X

matrix([[2.],
        [4.],
        [1.]])

bananas (b) = 2; oranges (o) = 4; papayas (p) = 1

In [31]:
print(A.dot(X))
print(B)

[[7.  ]
 [0.  ]
 [5.25]]
[[7.  ]
 [0.  ]
 [5.25]]


## Exercise 4

A landlord owns 3 houses, a 1-bedroom, a 2-bedroom , and a 3-bedroom house. 

The total rent he receives is 1240 dollars. 

He needs to make some repairs, where it costs 10% of the 1-bedroom house’s rent for its repairs, 20% of the 2-bedroom for its repairs, and 30% of the 3-bedroom house's rent for its repairs.  The total repair bill for all three houses was $276. 

The 3-bedroom house's rent is twice the 1-bedroom house’s rent. 

How much is the individual rent for three houses?

* b1 + b2 + b3 = 1240
* 0.1b1 + 0.2b2 + 0.3b3 = 276
* 2b1 = b3 => 2b1 + 0b2 - b3 = 0

In [33]:
# Create a system of equations
A = np.matrix([[1,1,1],[0.1,0.2,0.3],[2,0,-1]])
B = np.matrix([1240, 276, 0])
print(A)
print(B)

[[ 1.   1.   1. ]
 [ 0.1  0.2  0.3]
 [ 2.   0.  -1. ]]
[[1240  276    0]]


In [34]:
Ainv = np.linalg.inv(A)
print("A inverted", Ainv)

B = B.T
print("Matrix B", B)
print("Shape of B", B.shape)

X = Ainv.dot(B)
print("Matrix X", X)

A inverted [[ -2.  10.   1.]
 [  7. -30.  -2.]
 [ -4.  20.   1.]]
Matrix B [[1240]
 [ 276]
 [   0]]
Shape of B (3, 1)
Matrix X [[280.]
 [400.]
 [560.]]


* 1-bedroom house rent is 280 dollars, 
* 2-bedroom hpuse rent is 400 dollars, 
* 3-bedroom house rent is 560 dollars.

## Summary
In this lab, we saw how to use Numpy to solve linear equations by taking inverses and matrix multiplication and also using numpy's `.solve()`. We shall now take these skills forward and see how we can define a simple regression problem using linear algebra and solve it with numpy, only using the skills that we have learnt so far. 